### PR TITLE
feat: send credit notes with no refund to SAP

### PR DESCRIPTION
**Pull request by jhoffman-ms**
Allow to create a credit note with no refund  in SAP.

**Changes:**

- Added new endpoint to create a credit note in SAP.
- Add new UTs.

**Task:** [DAT-266](https://makingsense.atlassian.net/jira/software/projects/DAT/boards/62?selectedIssue=DAT-266)

### CODE REVIEW

File: Doppler.Sap.Test/BillingServiceTest.cs [hunk 1]
```diff
@@ -382,5 +382,146 @@ public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenUpdateBillingR
...
+
+            await billingService.CreateCreditNote(creditNote);
+
+            queuingServiceMock.Verify(x => x.AddToTaskQueue(It.IsAny<SapTask>()), Times.Once);
+        }
+
+        [Fact]
+        public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenCreateCreditNotesHasOneElementWithInvalidInvoiceId()
+        {
+            var sapConfigMock = new Mock<IOptions<SapConfig>>();
```
Comment (nreal) [inline Doppler.Sap.Test/BillingServiceTest.cs:425]: This variable is not used, you can remove it

File: Doppler.Sap.Test/BillingServiceTest.cs [hunk 2]
```diff
@@ -382,5 +382,146 @@ public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenUpdateBillingR
...
+        [Fact]
+        public async Task BillingService_ShouldBeAddOneTaskInQueue_WhenCreateCreditNotesHasOneValidElement()
+        {
+            var billingValidations = new List<IBillingValidation>
+            {
+                new BillingForArValidation(Mock.Of<ILogger<BillingForArValidation>>()),
+                new BillingForUsValidation(Mock.Of<ILogger<BillingForUsValidation>>())
+            };
+
+            var sapConfigMock = new Mock<IOptions<SapConfig>>();
```
Comment (nreal) [inline Doppler.Sap.Test/BillingServiceTest.cs:395]: This variable is not used, you can remove it

File: Doppler.Sap.Test/BillingServiceTest.cs [hunk 3]
```diff
@@ -382,5 +382,146 @@ public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenUpdateBillingR
...
+            await billingService.CreateCreditNote(creditNote);
+
+            queuingServiceMock.Verify(x => x.AddToTaskQueue(It.IsAny<SapTask>()), Times.Never);
+            slackServiceMock.Verify(x => x.SendNotification(It.IsAny<string>()), Times.Once);
+        }
+
+        [Fact]
+        public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenCreateCreditNotesHasOneElementWithInvalidBillingSystemId()
+        {
+            var sapConfigMock = new Mock<IOptions<SapConfig>>();
```
Comment (nreal) [inline Doppler.Sap.Test/BillingServiceTest.cs:460]: This variable is not used, you can remove it

File: Doppler.Sap.Test/BillingServiceTest.cs [hunk 4]
```diff
@@ -382,5 +382,146 @@ public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenUpdateBillingR
...
+            await billingService.CreateCreditNote(creditNote);
+
+            queuingServiceMock.Verify(x => x.AddToTaskQueue(It.IsAny<SapTask>()), Times.Never);
+            slackServiceMock.Verify(x => x.SendNotification(It.IsAny<string>()), Times.Once);
+        }
+
+        [Fact]
+        public async Task BillingService_ShouldBeNotAddOneTaskInQueue_WhenCreateCreditNotesHasOneElementWithInvalidClientId()
+        {
+            var sapConfigMock = new Mock<IOptions<SapConfig>>();
```
Comment (nreal) [inline Doppler.Sap.Test/BillingServiceTest.cs:495]: This variable is not used, you can remove it

File: Doppler.Sap.Test/CreditNoteHandlerTest.cs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        public async Task CreditNoteHandler_ShouldBeCreateCreditNoteInSap_WhenQueueHasOneValidElement()
+        {
+            var sapSaleOrderInvoiceResponse = new SapSaleOrderInvoiceResponse
+            {
+                BillingSystemId = 2,
+                CardCode = "CD001",
+                DocumentLines = new List<SapDocumentLineResponse>()
+            };
+
+            var billingValidations = new List<IBillingValidation>
```
Comment (nreal) [inline Doppler.Sap.Test/CreditNoteHandlerTest.cs:33]: This variable is not used, you can remove it

File: Doppler.Sap.Test/BillingServiceTest.cs (hunk 1 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/BillingServiceTest.cs]: Done

File: Doppler.Sap.Test/BillingServiceTest.cs (hunk 2 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/BillingServiceTest.cs]: Done

File: Doppler.Sap.Test/BillingServiceTest.cs (hunk 3 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/BillingServiceTest.cs]: Done

File: Doppler.Sap.Test/BillingServiceTest.cs (hunk 4 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/BillingServiceTest.cs]: Done

File: Doppler.Sap/Factory/CreditNoteHandler.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                    };
+                }
+            }
+            catch (Exception ex)
+            {
+                return new SapTaskResult
+                {
+                    IsSuccessful = false,
+                    SapResponseContent = ex.Message,
+                    TaskName = "Creating Creadit Note Request"
```
Comment (nreal) [inline Doppler.Sap/Factory/CreditNoteHandler.cs:74]: Very minor typo Cre**a**dit should be Credit

File: Doppler.Sap.Test/CreditNoteHandlerTest.cs (hunk 5 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/CreditNoteHandlerTest.cs]: Done

**Review by bamaro-dev — COMMENTED** [review]
🚀

**Comment by nreal** [comment]
💯

**Pull request merged**